compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp

```diff
@@ -464,6 +465,52 @@ FailureOr<Flow::DispatchRegionOp> Flow::wrapOpInDispatchRegion(
   return newRegionOp;
 }
 
+//===---------------------------------------------------------------------===//
+// UnCSE Clonable ops
+//===---------------------------------------------------------------------===//
+
+namespace {
+struct UnCseClonableOps : public RewritePattern {
+  UnCseClonableOps(MLIRContext *context)
+      : RewritePattern(MatchAnyOpTypeTag(), /*benefit=*/1, context) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    if (!isa<FunctionOpInterface>(op->getParentOp())) {
+      return rewriter.notifyMatchFailure(op, "parent is not a func.func");
+    }
+    if (!Flow::isClonableIntoDispatchOp(op)) {
+      return rewriter.notifyMatchFailure(op, "not clonable op");
+    }
+
+    if (llvm::hasSingleElement(op->getUsers())) {
+      return rewriter.notifyMatchFailure(op, "has single use");
+    }
+
+    OpBuilder::InsertionGuard g(rewriter);
+    auto users = llvm::to_vector(op->getUsers());
+    for (Operation *user : users) {
+      rewriter.setInsertionPoint(user);
+      Operation *clone = rewriter.clone(*op);
+      rewriter.replaceOpWithIf(op, clone->getResults(), [&](OpOperand &use) {
+        return use.getOwner() == user;
+      });
+    }
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+}  // namespace
+
+/// Clonabale operations with multiple uses can prevent maximal grouping of
+/// operations. Since all operations are tensor based, CSE can create false
+/// sharing. unCSE such operations.
+LogicalResult Flow::unCSEClonableOps(Operation *funcOp, MLIRContext *context) {
```

Comment:
This is a bit terrifying and something that risks compile time performance issues. It'd be really nice if we didn't require such an expensive step and could instead track this during the pass that needs it. I'm thinking of cases where this could easily create 10000+ ops in the larger models where we have 20-30k (which today are still only slices of models) at the stage where this may be run.

I was able to do a similar pass without needing to do this so it may be useful to dig into what particular thing this is helping with - may be able to have a set that accomplishes the same thing (build isCloneable Operation* set then lookup op in isCloneable set during formation).

Reply:
Removed all of this